File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        s_map, t_map = {}, {}
        for c in t:
            t_map[c] = t_map.get(c, 0) + 1
        
        cur, needed = 0, len(t_map)
        l, start = 0, 0
        min_length = len(s) + 1

        for r in range(len(s)):
            if s[r] not in t_map:
                continue
            
            s_map[s[r]] = s_map.get(s[r], 0) + 1
            if (s_map[s[r]] == t_map[s[r]]):
                cur += 1
                while cur == needed:
                    if r - l + 1 < min_length:
                        min_length = r - l + 1
                        start = l
                    
                    if s[l] in t_map:
                        s_map[s[l]] -= 1
                        if s_map[s[l]] < t_map[s[l]]:
                            cur -= 1
                    
                    l += 1
        
        return s[start: start + min_length] if min_length <= len(s) else ""

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        std::unordered_map<char, int> s_map;
        std::unordered_map<char, int> t_map;

        for (char c: t) {
            t_map[c]++;
        }

        int cur = 0, needed = t_map.size();
        int l = 0, start = 0;
        int min_length = s.length() + 1;

        for (int r = 0; r < s.length(); r++) {
            if (!t_map.count(s[r])) {
                continue;
            }
  
            s_map[s[r]]++;
            if (s_map[s[r]] == t_map[s[r]]) {
                cur++;
                while (cur == needed) {
                    if (r - l + 1 < min_length) {
                        min_length = r - l + 1;
                        start = l;
                    }

                    if (t_map.count(s[l])) {
                        s_map[s[l]]--;
                        if (s_map[s[l]] < t_map[s[l]]) {
                            cur--;
                        }
                    }

                    l++;
                }
            }
            
        }

        return min_length <= s.length() 
               ? s.substr(start, min_length) 
               : "";
    }
};

